### PR TITLE
📖 Add CRD Upgrade Safety documentation

### LINKED ISSUE #746: Docs: CRD Upgrade Safety
# User story

Describe the user story or stories for this epic:

- As a user, I expect updates to a CRD to be rejected (install/upgrade fails) when:
    - the scope changes from Cluster --> Namespace or Namespace --> Cluster
    - an existing stored version is removed
    - a new required field is added to an existing version
    - an existing field is removed from an existing version
    - an existing field type is changed in an existing version
    - a valid enum value is removed for a field in an existing version
- As a user, I expect updates to a CRD to be accepted (install/upgrade succeeds) when:
    - An enum value is added for a field in an existing version
    - An existing required field is updated to be optional in an existing version
    - The minimum value of an existing field is decreased in an existing version
    - The maximum value of an existing field is increased in an existing version
    - A new version of the CRD is added with no modifications to existing versions
- As a user I should be able to disable the CRD Upgrade Safety validations for a particular ClusterExtension
     
## List the user-facing issues in the epic that require documentation.
- #745
- #741
- Could probably be grouped into a singular section:
    - #662 
    - #663 
    - #664 
    - #665 
    - #714 
    - #715 
    - #716 
    - #713 

## Does this epic require new content or updates to existing content?

- [x] Net-new content
- [ ] Update existing content

## What type of documentation is required for this epic?

- [x] Release note
- [x] Concept
- [ ] Procedure
- [x] Reference

## Is there prior art or reference materials that would help write the documentation?
- Once the feature is complete and in place, there will likely be some unit and e2e tests that could be used as reference for how CRDs with allowed and disallowed changes look

Documentation should be contributed to the [`docs/drafts/` directory](https://github.com/operator-framework/operator-controller/tree/main/docs/drafts)

### PR DESCRIPTION
**Pull request by trgeiger**
# Description
Closes #746 

## Reviewer Checklist

- [ ] API Go Documentation
- [ ] Tests: Unit Tests (and E2E Tests, if appropriate)
- [ ] Comprehensive Commit Messages
- [ ] Links to related GitHub Issue(s)

### CODE REVIEW

File: docs/refs/crd-upgrade-safety.md
Comment (everettraven) [inline docs/refs/crd-upgrade-safety.md:24]:
```suggestion
- The minimum value of an existing field is increased in an existing version
```

File: docs/refs/crd-upgrade-safety.md
Comment (everettraven) [inline docs/refs/crd-upgrade-safety.md:36]: What do you think about moving this to being immediately after the list of validations?

File: docs/refs/crd-upgrade-safety.md
Comment (everettraven) [inline docs/refs/crd-upgrade-safety.md:33]: Should we include a statement along the lines of "If you identify any checks that we do not currently have coverage for and would like explicit coverage for, please create an issue"?

File: docs/refs/crd-upgrade-safety.md
Comment (everettraven) [inline docs/refs/crd-upgrade-safety.md:42]: I think it is fine to leave in for now because it is not considered a breaking change in our validations, but something we identified is that Kubernetes API conventions state that adding new enum values is actually considered a breaking change.

File: docs/refs/crd-upgrade-safety.md
Comment (everettraven) [inline docs/refs/crd-upgrade-safety.md:46]: No change needed in this PR, but something that I think we may have overlooked when originally developing the preflight checks is that new versions of the CRD must be backwards compatible with the served versions _OR_ provide conversion logic to prevent data loss and invalidation of resources stored at an older version.

@joelanford I think we may need to revisit the CRD Upgrade Safety preflight checks to ensure we capture behavior to check for breaking changes between all stored and served versions of the CRD when there is no conversion strategy specified.

File: docs/refs/crd-upgrade-safety.md
Comment (everettraven) [inline docs/refs/crd-upgrade-safety.md:136]: Should we include what the error message looks like when a change like this is caught?

File: docs/refs/crd-upgrade-safety.md
Comment (trgeiger) [reply docs/refs/crd-upgrade-safety.md:49]: given the long discussion this birthed in Slack, do we want to hold off on these docs until that work is done or would we just update this once that's complete? Not sure on what kind of timeline you were imagining all this.

File: docs/refs/crd-upgrade-safety.md
Comment (michaelryanpeter) [inline docs/refs/crd-upgrade-safety.md:71]: I might flip this around so that it is a little more direct:
```suggestion
You cannot individual field validators. If you disable the CRD Upgrade Safety preflight check, all the field validators are disabled.
```
Or something like that.

File: docs/refs/crd-upgrade-safety.md
Comment (michaelryanpeter) [reply docs/refs/crd-upgrade-safety.md:139]: +1

File: docs/refs/crd-upgrade-safety.md
Comment (michaelryanpeter) [inline docs/refs/crd-upgrade-safety.md:28]:
```suggestion
!!! note 
The rules for changes to minimum and maximum values apply to `minimum`, `minLength`,
```
See https://squidfunk.github.io/mkdocs-material/reference/admonitions/#usage

File: docs/refs/crd-upgrade-safety.md
Comment (michaelryanpeter) [inline docs/refs/crd-upgrade-safety.md:38]: Would we want to highlight this as an admonition to draw attention to it? Maybe as a tip?
https://squidfunk.github.io/mkdocs-material/reference/admonitions/#supported-types

File: docs/refs/crd-upgrade-safety.md
Comment (michaelryanpeter) [inline docs/refs/crd-upgrade-safety.md:73]: I think this admonition is more of a `warning`. WDYT?

File: docs/refs/crd-upgrade-safety.md
Comment (michaelryanpeter) [reply docs/refs/crd-upgrade-safety.md:139]: I like putting example outputs in [collapsible blocks](https://github.com/operator-framework/operator-controller/blob/main/docs/Tasks/explore-available-packages.md?plain=1#L25).

* https://squidfunk.github.io/mkdocs-material/reference/admonitions/#collapsible-blocks

**Review by michaelryanpeter — COMMENTED** [review]
Great work with very clear writing. I left a few suggestions. I noticed that this new page is not showing up in the Netlify preview.

File: docs/refs/crd-upgrade-safety.md
Comment (michaelryanpeter) [inline docs/refs/crd-upgrade-safety.md:12]: I think you might need a line break between the lead-in sentence and the unordered list:
image

File: docs/refs/crd-upgrade-safety.md
Comment (michaelryanpeter) [inline docs/refs/crd-upgrade-safety.md:44]: This list also needs an extra line break between the lead-in sentence and the list to fix the rendering in the preview:
image

**Review by michaelryanpeter — COMMENTED** [review]
A couple of nits to fix the rendering on the unordered lists in the preview. Otherwise, LGTM.

File: docs/refs/crd-upgrade-safety.md
Comment (trgeiger) [reply docs/refs/crd-upgrade-safety.md:24]: That could be nice. I spent a few minutes trying to think of how exactly that would be structured and couldn't really come up with a good idea. Maybe columns for safe/unsafe changes or something?

Comment (michaelryanpeter) [reply docs/refs/crd-upgrade-safety.md:24]: I have played around with it too, but I haven't been able to beat your list. I also haven't been able to figure out a good information architecture.


I do feel like like it would be nice to have this be a bit more scan-able as a reference. Maybe we will come up with something when we downstream this doc?

File: docs/refs/crd-upgrade-safety.md
Comment (michaelryanpeter) [inline docs/refs/crd-upgrade-safety.md:146]: I agree that the long line is a little hard to read, but I have always been told to keep example outputs as they are printed in the terminal.

**Review by michaelryanpeter — COMMENTED** [review]
lgtm (though I am not an owner/maintainer on the project)

File: docs/refs/crd-upgrade-safety.md
Comment (everettraven) [inline docs/refs/crd-upgrade-safety.md:28]:
```suggestion
    `minProperties`, `minItems`, `maximum`, `maxLength`, `maxProperties`, and `maxItems` constraints.
```

File: docs/refs/crd-upgrade-safety.md
Comment (everettraven) [reply docs/refs/crd-upgrade-safety.md:50]: We can let the doc through and update it as necessary

File: docs/refs/crd-upgrade-safety.md
Comment (everettraven) [inline docs/refs/crd-upgrade-safety.md:149]:
```suggestion
### Removing a stored version
```

File: docs/refs/crd-upgrade-safety.md
Comment (everettraven) [inline docs/refs/crd-upgrade-safety.md:151]:
```suggestion
In this example, the stored version `v1alpha1`, has been removed before it is no longer a stored version:
```

Comment (trgeiger) [reply docs/refs/crd-upgrade-safety.md:151]: The second clause of your suggestion is a little confusing to me. Is the distinction here that it's okay to remove a stored version if it's not in use? And not okay to move if it's being used? In which case would something like "has been removed while still in use?"

Comment (trgeiger) [reply docs/refs/crd-upgrade-safety.md:151]: Or maybe just simply "has been removed?"

**Review by everettraven — APPROVED** [review]

**Pull request merged**
